packages/synapse-sdk/src/storage/context.ts

```diff
@@ -615,48 +615,38 @@ export class StorageContext {
     )
 
     if (managedDataSets.length > 0 && !forceCreateDataSet) {
-      // Prefer data sets with pieces, sort by ID (older first)
-      const sorted = managedDataSets.sort((a, b) => {
-        if (a.currentPieceCount > 0 && b.currentPieceCount === 0) return -1
-        if (b.currentPieceCount > 0 && a.currentPieceCount === 0) return 1
-        return a.pdpVerifierDataSetId - b.pdpVerifierDataSetId
-      })
-
-      // Create async generator that yields providers lazily
-      async function* generateProviders(): AsyncGenerator<ProviderInfo> {
-        // First, yield providers from existing data sets (in sorted order)
-        for (const dataSet of sorted) {
-          if (skipProviderIds.has(dataSet.providerId)) {
-            continue
-          }
-          skipProviderIds.add(dataSet.providerId)
-          const provider = await spRegistry.getProvider(dataSet.providerId)
-
-          if (provider == null) {
-            console.warn(
-              `Provider ID ${dataSet.providerId} for data set ${dataSet.pdpVerifierDataSetId} is not currently approved`
-            )
-            continue
-          }
+      // Prefer data sets with pieces
+      const [hasNoPieces, hasPieces] = managedDataSets
+        .reduce<[Set<EnhancedDataSetInfo>, Set<EnhancedDataSetInfo>]>(
+          (results: [Set<EnhancedDataSetInfo>, Set<EnhancedDataSetInfo>], managedDataSet: EnhancedDataSetInfo) => {
+            results[managedDataSet.currentPieceCount > 0 ? 1 : 0].add(managedDataSet)
+            return results
+          },
+          [new Set(), new Set()]
+        )
+        .map((deduped) => [...deduped])
 
-          if (withIpni && provider.products.PDP?.data.ipniIpfs === false) {
-            continue
-          }
+      for (const managedDataSets of [hasPieces, hasNoPieces]) {
+        const providers: ProviderInfo[] = (
+          await Promise.all(
+            managedDataSets.map((dataSet: EnhancedDataSetInfo) => spRegistry.getProvider(dataSet.providerId))
```

Comment:
```suggestion
      for (const dataSets of [hasPieces, hasNoPieces]) {
        const providers: ProviderInfo[] = (
          await Promise.all(
            dataSets.map((dataSet: EnhancedDataSetInfo) => spRegistry.getProvider(dataSet.providerId))
```

shadowing `managedDataSets` here makes it confusing